Added infoscore command for dosage data

jigold opened the pull request:
- IMPUTE version of info score (http://www.nature.com/nrg/journal/v11/n7/extref/nrg2796-s3.pdf)
 - Sex chromosomes treated same as autosomes

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+  var nIncluded = 0
+
+  def expectedVariance(dosage: Array[Double]): Double = {
+    val mean = dosage(1) + 2 * dosage(2)
```
You also use this computation in `merge`, so maybe it should be abstracted?

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+import org.broadinstitute.hail.variant.Variant
+import org.broadinstitute.hail.variant._
+
+class InfoScoreCombiner extends Serializable {
```
Shouldn't need to be Serializable, since it's being called per variant

tpoterba replied:
also, I think this shouldn't be in driver.  How about the stats package?

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+      Some(1d - ((result / nIncluded) / (2 * theta * (1 - theta))))
+
+  def asAnnotation: Annotation = {
+    val imputeScore = thetaMLE.flatMap{theta => imputeInfoScore(theta)}
```
I think the current code style xml would put spaces around the brackets here.  Can you import those settings and autoformat again?

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+            assert(v == v2)
+            (v, (insertInfoScore(va, Some(comb.asAnnotation)), gs))
+          }
+        }.asOrderedRDD[Locus], vaSignature = newVAS)
```
don't need type bound [Locus] anymore

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+
+  def requiresVDS = true
+
+  val signature = TStruct(
```
put this signature on a companion object for DosageCombiner.  I also don't quite like the name "impute" but I don't have a suggestion for something better

tpoterba commented on src/main/scala/org/broadinstitute/hail/driver/InfoScore.scala:
```diff
@@ -0,0 +1,104 @@
+
+    val (newVAS, insertInfoScore) = vds.vaSignature.insert(signature, "infoscore")
+
+    state.copy(
```
should use VSM.mapAnnotationsWithAggregate instead of persisting & zipping, I think

tpoterba replied:
I think VariantQC should probably go in the same direction, until the query optimizer arrives.  (don't do this though, obviously!)

tpoterba replied:
However, do feel free to remove the unnecessary `combOp` asked for by `mapAnnotationsWithAggregate`.  That'll let you clean up your combiner by removing the second merge method as well

tpoterba commented on src/test/scala/org/broadinstitute/hail/methods/InfoScoreSuite.scala:
```diff
@@ -0,0 +1,54 @@
+    s = ImportGEN.run(s, Array("-s", sampleFile, genFile))
+    s = InfoScore.run(s, Array.empty[String])
+
+    val truthResult = sc.parallelize(hadoopConf.readLines(truthResultFile)(_.map(_.map { line =>
```
probably better to collect the hail result rather than parallelize the local result.

also, you could have achieved this by using `sc.readLines` rather than `sc.parallelize(hadoopConf.readLines(...`

jigold commented:
I moved the info score calculation to the stats package, got rid of the combiner class, and used `mapAnnotations` to modify the VDS. Let me know if I should use `mapAnnotationsWithAggregate` instead. I also removed the RDD join in the test suite.

The reason for not comparing to QCTOOL is how Cotton and I decided to handle missing values (0,0,0). QCTOOL uses the MAF estimate from the other samples to mean impute the dosage for samples with missing values. We do not include those samples at all in the computation.

tpoterba commented on src/main/scala/org/broadinstitute/hail/stats/package.scala:
```diff
@@ -2,9 +2,43 @@ package org.broadinstitute.hail
 
 package object stats {
 
+  def imputeInfoScore(gs: Iterable[Genotype]) = {
```
I kinda liked the combiner.  I see it having these advantages:

* can convert it to an annotation at the end with a class method
* can put the annotation schema on a companion object
* can have its merge function take Option[Array[Double]], which makes it totally general and somebody else could pick up that piece of code very easily
* supports aggregation rather than just folding, which helps to future-proof in case our representation ever distributes across samples as well, and follows the spark Combiner model

I would've put the combiner/companion object into a new file in the stats package.  That said, if you're happy with this structure, that's fine

tpoterba replied:
One thing, though: if we're going to keep `imputeInfoScore`, we should have its return type specified at declaration.  This makes code much more accessible

tpoterba commented on src/main/scala/org/broadinstitute/hail/stats/package.scala:
```diff
@@ -37,31 +71,39 @@ package object stats {
-          fa = fb; fb = fc; fc = fa
+        if (math.abs(fc) < math.abs(fb)) {
+          // swap endpoints so b is best approx
+          a = b;
```
either revert formatting change here or remove semicolons (probably the latter)

tpoterba commented:
We need a piece in the Aggregables docs with the other aggregators.  I think the best arrangement is a usage + examples section there, with the math and detail in the "Statistical functions" section.  They can link back and forth

tpoterba commented on src/main/scala/org/broadinstitute/hail/expr/AST.scala:
```diff
@@ -1187,7 +1187,17 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
+          case TGenotype => InfoScoreCombiner.signature
+          case _ => parseError(
+            s"""method `$method' can not operate on `$agg'
+                |  Accepted aggregable types: `Aggregable[Genotype]'
```
should be singular "type"

tpoterba commented on src/main/scala/org/broadinstitute/hail/stats/package.scala:
```diff
@@ -1,10 +1,63 @@
+
+    def merge(g: Genotype): InfoScoreCombiner = {
+      g.dosage.foreach { dx =>
+        val mean = expectedMean(dx)
```
expectedMean is now called only once, you can put the computation back inline I think.
